### PR TITLE
Updating http_network_or_cache_fetch to better match the fetch API spec

### LINKED ISSUE #14742: Update http_network_or_cache_fetch to current Fetch standard
http_network_or_cache_fetch looks to be out of date with what is currently in the [standard](https://fetch.spec.whatwg.org/#http-network-or-cache-fetch). It should be updated to reflect what the current standard contains.

### PR DESCRIPTION
**Pull request by mattnenterprise**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #14742

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14784)

### CODE REVIEW

**Comment by KiChjang** [comment]
```

  ▶ Unexpected subtest result in /XMLHttpRequest/send-entity-body-empty.htm:
  └ PASS [expected FAIL] XMLHttpRequest: send("") - empty entity body (HEAD)

  ▶ Unexpected subtest result in /XMLHttpRequest/send-entity-body-get-head.htm:
  └ PASS [expected FAIL] XMLHttpRequest: send() - non-empty data argument and GET/HEAD (HEAD)

  ▶ Unexpected subtest result in /XMLHttpRequest/send-entity-body-none.htm:
  └ PASS [expected FAIL] No content type (HEAD)

  ▶ Unexpected subtest result in /XMLHttpRequest/send-entity-body-none.htm:
  └ PASS [expected FAIL] Explicit content type (HEAD)

  ▶ Unexpected subtest result in /eventsource/request-cache-control.htm:
  └ PASS [expected TIMEOUT] EventSource: Cache-Control

  ▶ Unexpected subtest result in /html/webappapis/scripting/event-loops/microtask_after_script.html:
  └ PASS [expected FAIL] Microtask immediately after script
```
Looking good! You'll need to update the WPT test expectations.

File: tests/wpt/metadata/html/webappapis/scripting/event-loops/microtask_after_script.html.ini [hunk 1]
```diff
@@ -1,5 +0,0 @@
-[microtask_after_script.html]
-  type: testharness
-  [Microtask immediately after script]
-    expected: FAIL
-
```
Comment (jdm) [inline tests/wpt/metadata/html/webappapis/scripting/event-loops/microtask_after_script.html.ini]: Unfortunately this isn't actually a result caused by these changes (see #14221), so we should revert this.

**Comment by mattnenterprise** [comment]
I have made the updates.

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -967,25 +975,29 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
...
     } else if http_request.cache_mode.get() == CacheMode::Default ||
         http_request.cache_mode.get() == CacheMode::ForceCache {
         // TODO this substep
     }
 
-    // Step 18
+    // Step 20
     if response.is_none() {
+        if http_request.cache_mode.get() == CacheMode::OnlyIfCached {
+          return Response::network_error(NetworkError::Internal("Couldn't find response in cache".into()))
```
Comment (jdm) [inline components/net/http_loader.rs]: nit: four space indents, please.

File: components/net/http_loader.rs [hunk 3]
```diff
@@ -967,25 +975,29 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
...
     let response = response.unwrap();
 
-    // Step 19
     if let Some(status) = response.status {
-        if status == StatusCode::NotModified &&
-            (http_request.cache_mode.get() == CacheMode::Default ||
-            http_request.cache_mode.get() == CacheMode::NoCache) {
+      match status {
+        StatusCode::NotModified => {
+          // Step 21
```
Comment (jdm) [inline components/net/http_loader.rs]: nit: four space indents on this block and the previous line, please.

File: components/net/http_loader.rs [hunk 4]
```diff
@@ -967,25 +975,29 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
...
+          if http_request.cache_mode.get() == CacheMode::Default ||
+          http_request.cache_mode.get() == CacheMode::NoCache {
             // Substep 1
```
Comment (jdm) [inline components/net/http_loader.rs]: nit: this block should be indented four spaces past the previous line.

File: components/net/http_loader.rs [hunk 5]
```diff
@@ -1003,10 +1015,58 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
             // Substep 5
             // TODO cache_state is immutable?
             // response.cache_state = CacheState::Validated;
-        }
+          }
+        },
+        StatusCode::Unauthorized => {
+          // Step 22
```
Comment (jdm) [inline components/net/http_loader.rs]: nit: indent this block by another two spaces.

File: components/net/http_loader.rs [hunk 6]
```diff
@@ -1003,10 +1015,58 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
...
+              // otherwise this creates an infinite loop
+              // We basically pretend that the user declined to enter credentials
+              return response;
+          }
+
+          // Step 3
+          return http_network_or_cache_fetch(http_request, true, cors_flag, done_chan, context);
+        },
+        StatusCode::ProxyAuthenticationRequired => {
+          // Step 23
```
Comment (jdm) [inline components/net/http_loader.rs]: nit: indentation.

File: components/net/http_loader.rs [hunk 7]
```diff
@@ -1003,10 +1015,58 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
...
+          // return http_network_or_cache_fetch(request, authentication_fetch_flag, cors_flag, done_chan, context);
+        },
+        _ => {}
+      }
     }
 
-    // Step 20
+    // Step 24
+    if authentication_fetch_flag {
+      // TODO Create the authentication entry for request and the given realm
```
Comment (jdm) [inline components/net/http_loader.rs]: nit: indentation.

**Review by jdm — CHANGES_REQUESTED** [review]
This looks fine with some indentation adjustments.

**Comment by mattnenterprise** [comment]
I have made the updates.

File: components/net/http_loader.rs [hunk 8]
```diff
@@ -606,8 +607,8 @@ pub fn http_fetch(request: Rc<Request>,
         request.skip_service_worker.set(true);
 
         // Substep 3
-        let fetch_result = http_network_or_cache_fetch(request.clone(), credentials, authentication_fetch_flag,
-                                                       done_chan, context);
+        let fetch_result = http_network_or_cache_fetch(request.clone(), authentication_fetch_flag,
+          cors_flag, done_chan, context);
```
Comment (KiChjang) [inline components/net/http_loader.rs]: nit: Indentation.

File: components/net/http_loader.rs [hunk 9]
```diff
@@ -931,20 +938,21 @@ fn http_network_or_cache_fetch(request: Rc<Request>,
...
+    // Step 18
     // TODO have a HTTP cache to check for a completed response
     let complete_http_response_from_cache: Option<Response> = None;
     if http_request.cache_mode.get() != CacheMode::NoStore &&
         http_request.cache_mode.get() != CacheMode::Reload &&
         complete_http_response_from_cache.is_some() {
         // Substep 1
-        if http_request.cache_mode.get() == CacheMode::ForceCache {
+        if http_request.cache_mode.get() == CacheMode::ForceCache ||
+          http_request.cache_mode.get() == CacheMode::OnlyIfCached {
```
Comment (KiChjang) [inline components/net/http_loader.rs]: nit: Indentation, this should match the previous line's `http_request`.

**Review by KiChjang — CHANGES_REQUESTED** [review]
Almost there! Still needs some indentation changes, and after that it can be r=me.

**Comment by mattnenterprise** [comment]
I have made the updates.

**Comment by jdm** [comment]
@bors-servo r=KiChjang,jdm
Thanks!

**Comment by jdm** [comment]
```
Checking files for tidiness...
./components/net/http_loader.rs:1058: Line is longer than 120 characters
```

**Comment by mattnenterprise** [comment]
I have fixed that line.

**Comment by jdm** [comment]
@bors-servo: retry
* #13400

**Pull request merged**